### PR TITLE
[java] Support for annotations in types

### LINKED ISSUE #3752: [java] Expose annotations in symbol API
The class files and ASM give us access to annotations. The "hard" part of this feature, is that we need a way to represent annotation values symbolically. This adds significant API surface to the symbol API and it should probably stay `@Experimental` when PMD 7 is released.

Another complication is that the annotation values are easy to parse with ASM, but for symbols based on the AST, resolving those values requires name resolution. This introduces a dependency cycle between the name resolution and the symbol resolution pass. The cycle can be broken by resolving annotations lazily.

Branch: I think [this one](https://github.com/oowekyala/pmd/tree/typeres-annotations). I don't think ast-based annotations are supported yet in this branch.

This issue was split off from #2701

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

- Add type annotation to type mirrors
   - I modified the JTypeMirror interface and implementation to also carry around a set of type annotations.
   - This means no changes to the core inference are needed to propagate annotated types (in the simple cases I've tested yet).
- I changed the type of the method that returns annotations to return a `PSet<SymAnnot>` instead of `List<SymAnnot>`. We need a set because there is no defined iteration order for annotations that come from the class file, and we also have to merge annotations during inference. PSet is practical because we don't have to make copies to make the list immutable.
- Support parsing type annotation from ASM:
  - implemented for fields and methods/ctors, for type parameters and their bounds
  - not implemented for superclass/super interface signatures - can be done easily in another PR I think
- Same thing for AST types, the tests are shared between different implementations.

## Related issues

- Fixes #3752 
- Things to do later
	- support parsing annotations in supertype signatures (eg `class Foo extends List<@NonNull Integer>`)
	- verify that type inference merges type annotations properly when an ivar and instantiation have different type annotations

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/JClassSymbol.java [hunk 1]
```diff
@@ -247,6 +250,22 @@ default Set<String> getAnnotationAttributeNames() {
...
+     * given construct, as per the {@link Target} annotation. Return
+     * false if this is not an annotation.
+     */
+    default boolean annotationAppliesTo(ElementType elementType) {
+        if (!isAnnotation()) {
+            return false;
+        }
+        SymAnnot target = getDeclaredAnnotation(Target.class);
+        if (target == null) {
+            return false;
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/JClassSymbol.java:264]: This is not the right behavior… as per [the docs](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/Target.html):

> If an `@Target` meta-annotation is not present on an annotation type T , then an annotation of type T may be written as a modifier for any declaration except a type parameter declaration.

So, this should probably become:

```suggestion
        SymAnnot target = getDeclaredAnnotation(Target.class);
        if (target == null) {
            return elementType != ElementType.TYPE_PARAMETER;
```

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/JClassSymbol.java:264]: Thanks, I had no idea what the default was

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/JClassSymbol.java:264]: I should add unit tests for this as well

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 10 violations,<br />
introduces 1 new violations, 0 new errors and 0 new configuration errors,<br />
removes 1108 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/5e23dcdc597b48bea451cf723cf907e5/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 49509 violations,<br />
introduces 33904 new violations, 1446 new errors and 0 new configuration errors,<br />
removes 196500 violations, 4 errors and 7 configuration errors.<br />
<a href="https://chunk.io/pmd/5e23dcdc597b48bea451cf723cf907e5/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/MethodInfoVisitor.java [hunk 2]
```diff
@@ -22,18 +23,34 @@ class MethodInfoVisitor extends MethodVisitor {
...
+            public void visitEnd() {
+                execStub.addParameterAnnotation(parameter, annot);
+            }
+        };
     }
 
     @Override
     public void visitEnd() {
         execStub.setDefaultAnnotValue(defaultAnnotValue);
-        execStub.finalizeVisit();
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/MethodInfoVisitor.java:36]: method was removed

**Comment by jsotuyod** [comment]
@oowekyala is this now ready for review?

**Comment by jsotuyod** [comment]
This PR states that there is future work:

> Things to do later
> * support parsing annotations in supertype signatures (eg `class Foo extends List<@NonNull Integer>`)
> * verify that type inference merges type annotations properly when an ivar and instantiation have different type annotations

I'd add to this list that return types for constructors are currently not including annotations.

Given the code (present in this PR):

```java
public abstract class ClassWithTypeAnnotationsOnMethods {
     static class CtorOwner {

        CtorOwner(@A @B int i) { }

        @A CtorOwner() { }

        CtorOwner(String i) throws @A Exception {}
    }
}
```

The following test would fail:

```java
JClassType symParent = impl.getDeclaration(ClassWithTypeAnnotationsOnMethods.class);
JClassType sym = symParent.getDeclaredClass("CtorOwner");
assertHasTypeAnnots(sym.getConstructors().get(1).getReturnType(), ANNOT_A); // FAILS!
```

**Pull request merged**